components/script/script_module.rs

```diff
@@ -882,26 +928,53 @@ impl Callback for ModuleHandler {
     }
 }
 
+/// An unique id for dynamic module
+#[derive(Clone, Copy, Debug, Eq, Hash, JSTraceable, PartialEq)]
+pub(crate) struct DynamicModuleId(Uuid);
+
+/// Owner of dynamic module
+#[dom_struct]
+pub(crate) struct DynamicModuleOwner {
+    #[ignore_malloc_size_of = "Rc"]
+    promise: Rc<Promise>,
+
+    /// Due to different from an usual script element, we will need a
+    /// standalone owner for dynamic modules.
+    /// The owner will hold the dynamic module's id and then it can be used
+    /// to get the dynamic module from `dynamic_module_list` from global scope.
+    #[ignore_malloc_size_of = "Defined in uuid"]
+    module_id: DynamicModuleId,
+}
+
+impl DynamicModuleOwner {
+    fn new(promise: Rc<Promise>, module_id: DynamicModuleId) -> DynamicModuleOwner {
+        DynamicModuleOwner { promise, module_id }
+    }
+}
+
 /// The owner of the module
 /// It can be `worker` or `script` element
 #[derive(Clone)]
-pub enum ModuleOwner {
+pub(crate) enum ModuleOwner {
     #[allow(dead_code)]
     Worker(TrustedWorkerAddress),
     Window(Trusted<HTMLScriptElement>),
+    DynamicModule(Trusted<DynamicModuleOwner>),
```

Comment:
I can't figure out how this compiles, given `pub struct Trusted<T: DomObject>` in dom/bindings/refcounted.rs. I am particularly confused because I am clearly compiling your branch that has this line in it!

Reply:
Because the first field of `DynamicModuleOwner` is `promise` which can be a trusted DomObject 👀 

https://github.com/servo/servo/blob/d5808e0154cc526909799fc4134ba742a3dab906/components/script/script_module.rs#L935-L939

Is this a valid usage of Trusted refcounted thing?

(btw, the `finish_dynamic_module` will be moved to `DynamicModuleOwner` arm in `notify_owner_to_finish` but I was busy this week >_< so haven't tried to fix it; also, I'm wondering it might be good to have a trait to define the `notify_owner_to_finish` function? so that we can just call the function in `ModuleOwner::notify_owner_to_finish`)

Reply:
Oh, I completely overlooked the `#[dom_struct]` on DynamicModuleOwner. We require `TrustedPromise` instead of `Trusted<Promise>`, and we should consider some other way of holding the DynamicModuleOwner. I honestly don't know if the current implementation is expected to work or not.

Reply:
I just figured out why we're hitting TIMEOUT for some tests and why https://austineng.github.io/webgpu-samples/ also gets the TIMEOUT.

Basically, the root cause is,

when the owner is `ModuleOwner::Script`, we will send the IPC route to `CoreResourceMsg::Fetch` via the `document_loader::fetch_async`; however, when the owner is a dynamic loader (or a worker in the future), nothing will handle the IPC router so it will just be dropped on the floor. Then, no one knows the route should be sent.

https://github.com/servo/servo/blob/d5808e0154cc526909799fc4134ba742a3dab906/components/script/script_module.rs#L1708-L1717

So, we should handle the IPC router much correctly for non-script cases.

then, while I tried to experimentally send the IPC action

```rust
match document {
    Some(doc) => doc.fetch_async(LoadType::Script(url.clone()), request, action_sender),
    None => {
        let _ = global
            .resource_threads()
            .sender()
            .send(CoreResourceMsg::Fetch(
                request,
                FetchChannels::ResponseMsg(action_sender, None),
            ))
            .unwrap();
    },
}
```

> I honestly don't know if the current implementation is expected to work or not.

I just got the answer of this and it's `No, it won't work`.

Reply:
Well spotted!